L1Trigger/TrackTrigger/src/TrackQuality.cc

```diff
@@ -0,0 +1,298 @@
+  }
+
+  int tmp_trk_dtot = 0;
+  //calculate number of disk hits
+  for (int i = 6; i < 11; ++i) {
+    tmp_trk_dtot += hitpattern_expanded_binary[i];
+  }
+
+  // bin bendchi2 variable (bins from https://twiki.cern.ch/twiki/bin/viewauth/CMS/HybridDataFormat#Fitted_Tracks_written_by_KalmanF)
+  float tmp_trk_bendchi2 = aTrack.stubPtConsistency();
+  std::vector<float> bendchi2_bins = {0, 0.5, 1.25, 2, 3, 5, 10, 50};
+  int n_bendchi2 = static_cast<int>(bendchi2_bins.size());
+  float tmp_trk_bendchi2_bin = -1;
+  for (int i = 0; i < n_bendchi2; i++) {
+    if (tmp_trk_bendchi2 >= bendchi2_bins[i] && tmp_trk_bendchi2 < bendchi2_bins[i + 1]) {
+      tmp_trk_bendchi2_bin = i;
+      break;
+    }
+  }
+  if (tmp_trk_bendchi2_bin < 0)
+    tmp_trk_bendchi2_bin = n_bendchi2;
+
+  // bin chi2rphi variable (bins from https://twiki.cern.ch/twiki/bin/viewauth/CMS/HybridDataFormat#Fitted_Tracks_written_by_KalmanF)
+  float tmp_trk_chi2rphi = aTrack.chi2XY();
+  std::vector<float> chi2rphi_bins = {0, 0.25, 0.5, 1, 2, 3, 5, 7, 10, 20, 40, 100, 200, 500, 1000, 3000};
+  int n_chi2rphi = static_cast<int>(chi2rphi_bins.size());
+  float tmp_trk_chi2rphi_bin = -1;
+  for (int i = 0; i < n_chi2rphi; i++) {
+    if (tmp_trk_chi2rphi >= chi2rphi_bins[i] && tmp_trk_chi2rphi < chi2rphi_bins[i + 1]) {
+      tmp_trk_chi2rphi_bin = i;
+      break;
+    }
+  }
+  if (tmp_trk_chi2rphi_bin < 0)
+    tmp_trk_chi2rphi_bin = n_chi2rphi;
+
+  // bin chi2rz variable (bins from https://twiki.cern.ch/twiki/bin/viewauth/CMS/HybridDataFormat#Fitted_Tracks_written_by_KalmanF)
+  float tmp_trk_chi2rz = aTrack.chi2Z();
+  std::vector<float> chi2rz_bins = {0, 0.25, 0.5, 1, 2, 3, 5, 7, 10, 20, 40, 100, 200, 500, 1000, 3000};
+  int n_chi2rz = static_cast<int>(chi2rz_bins.size());
+  float tmp_trk_chi2rz_bin = -1;
+  for (int i = 0; i < n_chi2rz; i++) {
+    if (tmp_trk_chi2rz >= chi2rz_bins[i] && tmp_trk_chi2rz < chi2rz_bins[i + 1]) {
+      tmp_trk_chi2rz_bin = i;
+      break;
+    }
+  }
+  if (tmp_trk_chi2rz_bin < 0)
+    tmp_trk_chi2rz_bin = n_chi2rz;
+
+  // get the nstub
+  std::vector<edm::Ref<edmNew::DetSetVector<TTStub<Ref_Phase2TrackerDigi_>>, TTStub<Ref_Phase2TrackerDigi_>>> stubRefs =
+      aTrack.getStubRefs();
+  float tmp_trk_nstub = stubRefs.size();
+
+  // While not strictly necessary to define these parameters,
+  // it is included so each variable is named to avoid confusion
+  float tmp_trk_big_invr = 500 * abs(aTrack.rInv());
+  float tmp_trk_tanl = abs(aTrack.tanL());
+  float tmp_trk_z0 = aTrack.z0();
+  float tmp_trk_phi = aTrack.phi();
+  float tmp_trk_pt = aTrack.momentum().perp();
+  float tmp_trk_eta = aTrack.eta();
+  float tmp_trk_chi2 = aTrack.chi2();
+  float tmp_trk_log_chi2 = log(tmp_trk_chi2);
+  float tmp_trk_log_chi2rphi = log(tmp_trk_chi2rphi);
+  float tmp_trk_log_chi2rz = log(tmp_trk_chi2rz);
+  float tmp_trk_log_bendchi2 = log(tmp_trk_bendchi2);
+
+  // fill feature map
+  std::map<std::string, float> feature_map;
```

Comment:
by declaring and using this at the beginning, many temporaries could be avoided

Reply:
done (some temporaries remain due to how they are calculated however)